### PR TITLE
Implement expand module intention

### PR DESCRIPTION
**Pull request by pavel-v-chernykh**
Implement expand module intention

### CODE REVIEW

**Comment by pavel-v-chernykh** [comment]
@matklad please, take a look, it is related to #123

**Comment by matklad** [comment]
@pavel-v-chernykh can you add a test?

**Comment by pavel-v-chernykh** [comment]
@matklad Could you provide some hint how to do it?

**Comment by matklad** [comment]
Here is something similar: https://upsource.jetbrains.com/idea-community/file/995e5a44397522c9625ba66dc7efbe534c48499e/python%2FtestSrc%2Fcom%2Fjetbrains%2Fpython%2Frefactoring%2FPyMoveTest.java?nav=10493:10507:focused&line=474

File: src/org/rust/lang/intentions/ExpandModule.kt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+    override fun isAvailable(project: Project, editor: Editor?, file: PsiFile?): Boolean {
+        return file is RustFileImpl && file.name != "mod.rs"
+    }
+
+    override fun invoke(project: Project, editor: Editor?, file: PsiFile?) {
+        val dirName = FileUtil.getNameWithoutExtension(file?.name as String)
+        val directory = file?.parent?.createSubdirectory(dirName)
+        MoveFilesOrDirectoriesUtil.doMoveFile(file, directory)
+        PsiFileImplUtil.setName(file as PsiFile, "mod.rs")
```
Comment (matklad) [inline src/org/rust/lang/intentions/ExpandModule.kt]: This casts look suspiciously and are asking for an NPE. The safe strategy is to check for nulls at the beginning and do an early return, like this

```Kotlin
val name = file?.name ?: return
```

**Comment by pavel-v-chernykh** [comment]
@matklad Please, take a look

File: src/org/rust/lang/intentions/ExpandModule.kt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        return "Expand module structure"
+    }
+
+    override fun startInWriteAction(): Boolean {
+        return true
+    }
+
+    override fun getText(): String {
+        return "Expand Module"
+    }
```
Comment (matklad) [inline src/org/rust/lang/intentions/ExpandModule.kt]: expressions bodies are usefull for such cases
```Kotlin
override fun getFamilyName() = "Expand..."

override fun startInWriteAction() = true
```

File: src/org/rust/lang/intentions/ExpandModule.kt [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    override fun isAvailable(project: Project, editor: Editor?, file: PsiFile?): Boolean {
+        return file is RustFileImpl && file.name != "mod.rs"
+    }
+
+    override fun invoke(project: Project, editor: Editor?, file: PsiFile?) {
+        val fileName = file?.name ?: return
+        val dirName = FileUtil.getNameWithoutExtension(fileName)
+        val directory = file?.parent?.createSubdirectory(dirName)
+        MoveFilesOrDirectoriesUtil.doMoveFile(file, directory)
+        PsiFileImplUtil.setName(file as PsiFile, "mod.rs")
```
Comment (matklad) [inline src/org/rust/lang/intentions/ExpandModule.kt]: This still can fail with NPE, if I understand `as` semantics correctly.

```Kotlin
if (file == null) {
    return
}
```

at the start of fun should help.

File: test/org/rust/lang/intentions/RustIntentionsTest.kt [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+import com.intellij.openapi.vfs.LocalFileSystem
+import com.intellij.testFramework.PlatformTestUtil
+import org.assertj.core.api.Assertions
+import org.rust.lang.RustTestCase
+import java.io.File
+
+class RustIntentionsTest : RustTestCase() {
+    override fun getTestDataPath() = "testData/org/rust/lang/intentions/fixtures/"
+
+    private fun doTest(action: IntentionAction) {
```
Comment (matklad) [inline test/org/rust/lang/intentions/RustIntentionsTest.kt]: What is the purpose of `doTest` function?

I guess you can simply write

```Kotlin
fun testExpandModule() {
//body of doTest
}

```

File: src/org/rust/lang/intentions/ExpandModule.kt (hunk 3 above)
Comment (pavel-v-chernykh) [inline src/org/rust/lang/intentions/ExpandModule.kt]: If file is null then function returns at first line ```val fileName = file?.name ?: return```

**Comment by pavel-v-chernykh** [comment]
@matklad one more try :smile:

File: .gitignore [hunk 5]
```diff
@@ -1,6 +1,7 @@
 .idea
 *.iml
 .gradle
+.sandbox
```
Comment (Turbo87) [inline .gitignore]: :+1:

File: resources/intentionDescriptions/ExpandModule/description.html [hunk 6]
```diff
@@ -0,0 +1,5 @@
+<html>
+<body>
+This intention expand your file based module into directory based module.
```
Comment (Turbo87) [inline resources/intentionDescriptions/ExpandModule/description.html]: "This intention expands your file based module into a directory based module."

File: .gitignore (hunk 5 above)
Comment (pavel-v-chernykh) [inline .gitignore]: :smirk:

File: src/org/rust/lang/intentions/ExpandModule.kt (hunk 3 above)
Comment (matklad) [inline src/org/rust/lang/intentions/ExpandModule.kt]: I see, but, unfortunately, Kotlin's flow sensitive type inference does not. 

You should use `file!!` to cast away `?`. I didn't know that `as` allows to cast away `?` btw ;)

Still, I would prefer an explicit check at the beginning of the function: invariant checked by the compiler tend to be more invariant.

File: resources/intentionDescriptions/ExpandModule/description.html (hunk 6 above)
Comment (pavel-v-chernykh) [inline resources/intentionDescriptions/ExpandModule/description.html]: @Turbo87 Good catch. Fixed

File: src/org/rust/lang/intentions/ExpandModule.kt (hunk 3 above)
Comment (pavel-v-chernykh) [inline src/org/rust/lang/intentions/ExpandModule.kt]: @matklad added explicit ```file != null```, please check it

File: src/org/rust/lang/intentions/ExpandModule.kt [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    private val MOD_FILE_NAME = "mod.rs"
+
+    override fun getFamilyName() = "Expand module structure"
+    override fun startInWriteAction() = true
+    override fun getText() = "Expand Module"
+    override fun isAvailable(project: Project, editor: Editor?, file: PsiFile?) =
+        file is RustFileImpl && file.name != MOD_FILE_NAME
+
+    override fun invoke(project: Project, editor: Editor?, file: PsiFile?) {
+        if (file != null) {
```
Comment (matklad) [inline src/org/rust/lang/intentions/ExpandModule.kt]: I would suggest to save an indent level here: 

```Kotlin
file ?: return

val dirName = ...

```

**Comment by matklad** [comment]
:+1: from me, @atsky please take a look.

**Pull request merged**